gluten-substrait/src/main/java/org/apache/gluten/utils/SparkToJavaConverter.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gluten.utils;
+
+import org.apache.gluten.substrait.type.BinaryTypeNode;
+import org.apache.gluten.substrait.type.BooleanTypeNode;
+import org.apache.gluten.substrait.type.DateTypeNode;
+import org.apache.gluten.substrait.type.DecimalTypeNode;
+import org.apache.gluten.substrait.type.FP32TypeNode;
+import org.apache.gluten.substrait.type.FP64TypeNode;
+import org.apache.gluten.substrait.type.I16TypeNode;
+import org.apache.gluten.substrait.type.I32TypeNode;
+import org.apache.gluten.substrait.type.I64TypeNode;
+import org.apache.gluten.substrait.type.I8TypeNode;
+import org.apache.gluten.substrait.type.ListNode;
+import org.apache.gluten.substrait.type.MapNode;
+import org.apache.gluten.substrait.type.StringTypeNode;
+import org.apache.gluten.substrait.type.StructNode;
+import org.apache.gluten.substrait.type.TimestampTypeNode;
+import org.apache.gluten.substrait.type.TypeNode;
+
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.UnsafeArrayData;
+import org.apache.spark.sql.catalyst.util.ArrayData;
+import org.apache.spark.sql.catalyst.util.MapData;
+import org.apache.spark.sql.types.ArrayType;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.Decimal;
+import org.apache.spark.sql.types.MapType;
+import org.apache.spark.sql.types.StructType;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class SparkToJavaConverter {
+
+  private static Object toJava(Object obj, TypeNode nodeType) {
+    return toJava(obj, nodeType, null);
+  }
+
+  public static Object toJava(Object obj, TypeNode nodeType, DataType dataType) {
```

Comment:
Signature of the API seems like too general? `Object` used in both parameter list and return type. Is there a chance we can use strong types? Can we leverage visitor pattern?

Reply:
@shuai-xu Could check this https://github.com/apache/spark/blob/e428fe902bb1f12cea973de7fe4b885ae69fd6ca/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala#L489-L500

Reply:
This class also has a same signature `def convertToScala(catalystValue: Any, dataType: DataType): Any = `, it can not be used in gluten, or else will throw exceptions. And it return a spark Row for StructType.